--- Auth/Yadis/ParanoidHTTPFetcher.php	2012-10-21 00:26:52.036842042 +0200
+++ Auth/Yadis/ParanoidHTTPFetcher.php	2012-10-21 00:33:46.515660341 +0200
@@ -142,6 +142,19 @@
                 }
             }
 
+			// <TYPO3-specific>
+			if ($GLOBALS['TYPO3_CONF_VARS']['SYS']['curlProxyServer']) {
+				curl_setopt($c, CURLOPT_PROXY, $GLOBALS['TYPO3_CONF_VARS']['SYS']['curlProxyServer']);
+
+				if ($GLOBALS['TYPO3_CONF_VARS']['SYS']['curlProxyTunnel']) {
+					curl_setopt($c, CURLOPT_HTTPPROXYTUNNEL, $GLOBALS['TYPO3_CONF_VARS']['SYS']['curlProxyTunnel']);
+				}
+				if ($GLOBALS['TYPO3_CONF_VARS']['SYS']['curlProxyUserPass']) {
+					curl_setopt($c, CURLOPT_PROXYUSERPWD, $GLOBALS['TYPO3_CONF_VARS']['SYS']['curlProxyUserPass']);
+				}
+			}
+			// </TYPO3-specific>
+
             curl_exec($c);
 
             $code = curl_getinfo($c, CURLINFO_HTTP_CODE);
@@ -226,6 +239,19 @@
             }
         }
 
+		// <TYPO3-specific>
+		if ($GLOBALS['TYPO3_CONF_VARS']['SYS']['curlProxyServer']) {
+			curl_setopt($c, CURLOPT_PROXY, $GLOBALS['TYPO3_CONF_VARS']['SYS']['curlProxyServer']);
+
+			if ($GLOBALS['TYPO3_CONF_VARS']['SYS']['curlProxyTunnel']) {
+				curl_setopt($c, CURLOPT_HTTPPROXYTUNNEL, $GLOBALS['TYPO3_CONF_VARS']['SYS']['curlProxyTunnel']);
+			}
+			if ($GLOBALS['TYPO3_CONF_VARS']['SYS']['curlProxyUserPass']) {
+				curl_setopt($c, CURLOPT_PROXYUSERPWD, $GLOBALS['TYPO3_CONF_VARS']['SYS']['curlProxyUserPass']);
+			}
+		}
+		// </TYPO3-specific>
+
         curl_exec($c);
 
         $code = curl_getinfo($c, CURLINFO_HTTP_CODE);
