diff --git a/node_modules/@ckeditor/ckeditor5-utils/dist/dom/emittermixin.d.ts b/node_modules/@ckeditor/ckeditor5-utils/dist/dom/emittermixin.d.ts
index caf9542..5b34b9e 100644
--- a/node_modules/@ckeditor/ckeditor5-utils/dist/dom/emittermixin.d.ts
+++ b/node_modules/@ckeditor/ckeditor5-utils/dist/dom/emittermixin.d.ts
@@ -17,10 +17,10 @@ import type { Constructor, Mixed } from "../mix.js";
 * class MyDomEmitter extends MyDomEmitterBase {}
 * ```
 */
-export type DomEmitterMixinConstructor<Base extends Constructor<Emitter> | undefined = undefined> = Base extends Constructor<Emitter> ? Mixed<Base, DomEmitter> : {
+export type DomEmitterMixinConstructor<Base extends Constructor<Emitter> | undefined = undefined> = Base extends undefined ? {
 	new (): DomEmitter;
 	prototype: DomEmitter;
-};
+} : Mixed<Base, DomEmitter>;
 /**
 * Mixin that injects the DOM events API into its host. It provides the API
 * compatible with {@link module:utils/emittermixin~Emitter}.
diff --git a/node_modules/@ckeditor/ckeditor5-utils/dist/emittermixin.d.ts b/node_modules/@ckeditor/ckeditor5-utils/dist/emittermixin.d.ts
index b175cd7..e52fb21 100644
--- a/node_modules/@ckeditor/ckeditor5-utils/dist/emittermixin.d.ts
+++ b/node_modules/@ckeditor/ckeditor5-utils/dist/emittermixin.d.ts
@@ -18,10 +18,10 @@ import "./version.js";
 * class MyEmitter extends MyEmitterBase {}
 * ```
 */
-export type EmitterMixinConstructor<Base extends Constructor | undefined = undefined> = Base extends Constructor ? Mixed<Base, Emitter> : {
+export type EmitterMixinConstructor<Base extends Constructor | undefined = undefined> = Base extends undefined ? {
 	new (): Emitter;
 	prototype: Emitter;
-};
+} : Mixed<Base, Emitter>;
 /**
 * Mixin that injects the {@link ~Emitter events API} into its host.
 *
diff --git a/node_modules/@ckeditor/ckeditor5-utils/dist/observablemixin.d.ts b/node_modules/@ckeditor/ckeditor5-utils/dist/observablemixin.d.ts
index 848550c..e40f881 100644
--- a/node_modules/@ckeditor/ckeditor5-utils/dist/observablemixin.d.ts
+++ b/node_modules/@ckeditor/ckeditor5-utils/dist/observablemixin.d.ts
@@ -16,10 +16,10 @@ import type { Constructor, Mixed } from "./mix.js";
 * class MyObservable extends MyObservableBase {}
 * ```
 */
-export type ObservableMixinConstructor<Base extends Constructor<Emitter> | undefined = undefined> = Base extends Constructor<Emitter> ? Mixed<Base, Observable> : {
+export type ObservableMixinConstructor<Base extends Constructor<Emitter> | undefined = undefined> = Base extends undefined ? {
 	new (): Observable;
 	prototype: Observable;
-};
+} : Mixed<Base, Observable>;
 /**
 * A mixin that injects the "observable properties" and data binding functionality described in the
 * {@link ~Observable} interface.
