diff --git a/src/Exceptions/NestingException.php b/src/Exceptions/NestingException.php
index 474b987..7acc842 100644
--- a/src/Exceptions/NestingException.php
+++ b/src/Exceptions/NestingException.php
@@ -18,7 +18,7 @@ class NestingException extends \Exception
      * @param Exception|null   $previous
      * @param \DOMElement|null $element
      */
-    public function __construct($message = "", $code = 0, Exception $previous = null, \DOMElement $element = null)
+    public function __construct($message = "", $code = 0, ?Exception $previous = null, ?\DOMElement $element = null)
     {
         $this->element = $element;
         parent::__construct($message, $code, $previous);
