diff --git a/node_modules/devbridge-autocomplete/dist/jquery.autocomplete.js b/node_modules/devbridge-autocomplete/dist/jquery.autocomplete.js
index 74121fa..50d1494 100644
--- a/node_modules/devbridge-autocomplete/dist/jquery.autocomplete.js
+++ b/node_modules/devbridge-autocomplete/dist/jquery.autocomplete.js
@@ -667,7 +667,8 @@
                 beforeRender = options.beforeRender,
                 html = '',
                 category,
-                formatGroup = function (suggestion, index) {
+                // modified for TYPO3, added index parameter
+                formatGroup = function (suggestion, value, index) {
                         var currentCategory = suggestion.data[groupBy];

                         if (category === currentCategory){
@@ -676,7 +677,8 @@

                         category = currentCategory;

-                        return options.formatGroup(suggestion, category);
+                        // modified for TYPO3, added index parameter
+                        return options.formatGroup(suggestion, category, index);
                     };

             if (options.triggerSelectOnValidInput && that.isExactMatch(value)) {
