diff -ruw /home/karsten/Desktop/adodb/adodb-lib.inc.php adodb/adodb-lib.inc.php
--- /home/karsten/Desktop/adodb/adodb-lib.inc.php	2006-01-25 04:58:25.000000000 +0100
+++ adodb/adodb-lib.inc.php	2006-02-16 00:20:48.000000000 +0100
@@ -778,7 +778,7 @@
 	$values = substr($values, 0, -2);
 
 	// Append the fields and their values to the insert query.
-	return 'INSERT INTO '.$tableName.' ( '.$fields.' ) VALUES ( '.$values.' )';
+	return 'INSERT INTO '.$zthis->nameQuote.$tableName.$zthis->nameQuote.' ( '.$fields.' ) VALUES ( '.$values.' )';
 }
 
 
diff -ruw /home/karsten/Desktop/adodb/datadict/datadict-oci8.inc.php adodb/datadict/datadict-oci8.inc.php
--- /home/karsten/Desktop/adodb/datadict/datadict-oci8.inc.php	2006-01-24 05:37:21.000000000 +0100
+++ adodb/datadict/datadict-oci8.inc.php	2006-02-16 00:20:48.000000000 +0100
@@ -251,8 +251,8 @@
 		}
 		
 		if ( is_array($flds) )
-			$flds = implode(', ',$flds);
-		$s = 'CREATE' . $unique . ' INDEX ' . $idxname . ' ON ' . $tabname . ' (' . $flds . ')';
+			$flds = implode($this->quote.', '.$this->quote,$flds);
+		$s = 'CREATE' . $unique . ' INDEX ' .$this->quote. $idxname .$this->quote. ' ON ' . $this->quote.$tabname . $this->quote.' (' .$this->quote. $flds .$this->quote. ')';
 		
 		if ( isset($idxoptions[$this->upperName]) )
 			$s .= $idxoptions[$this->upperName];
diff -ruw /home/karsten/Desktop/adodb/datadict/datadict-postgres.inc.php adodb/datadict/datadict-postgres.inc.php
--- /home/karsten/Desktop/adodb/datadict/datadict-postgres.inc.php	2006-01-24 05:37:21.000000000 +0100
+++ adodb/datadict/datadict-postgres.inc.php	2006-02-16 00:20:48.000000000 +0100
@@ -342,7 +342,7 @@
 		
 		$unique = isset($idxoptions['UNIQUE']) ? ' UNIQUE' : '';
 		
-		$s = 'CREATE' . $unique . ' INDEX ' . $idxname . ' ON ' . $tabname . ' ';
+		$s = 'CREATE' . $unique . ' INDEX "' . $idxname . '" ON "' . $tabname . '" ';
 		
 		if (isset($idxoptions['HASH']))
 			$s .= 'USING HASH ';
@@ -351,8 +351,8 @@
 			$s .= $idxoptions[$this->upperName];
 		
 		if ( is_array($flds) )
-			$flds = implode(', ',$flds);
-		$s .= '(' . $flds . ')';
+			$flds = implode('", "',$flds);
+		$s .= '("' . $flds . '")';
 		$sql[] = $s;
 		
 		return $sql;
diff -ruw /home/karsten/Desktop/adodb/drivers/adodb-oci8.inc.php adodb/drivers/adodb-oci8.inc.php
--- /home/karsten/Desktop/adodb/drivers/adodb-oci8.inc.php	2006-01-24 05:37:31.000000000 +0100
+++ adodb/drivers/adodb-oci8.inc.php	2006-02-16 00:20:48.000000000 +0100
@@ -104,7 +104,7 @@
 		$ADODB_FETCH_MODE = ADODB_FETCH_NUM;
 		if ($this->fetchMode !== false) $savem = $this->SetFetchMode(false);
 		
-		$rs = $this->Execute(sprintf($this->metaColumnsSQL,strtoupper($table)));
+		$rs = $this->Execute(sprintf($this->metaColumnsSQL,$table));
 		
 		if (isset($savem)) $this->SetFetchMode($savem);
 		$ADODB_FETCH_MODE = $save;
diff -ruw /home/karsten/Desktop/adodb/drivers/adodb-postgres64.inc.php adodb/drivers/adodb-postgres64.inc.php
--- /home/karsten/Desktop/adodb/drivers/adodb-postgres64.inc.php	2006-01-24 05:37:31.000000000 +0100
+++ adodb/drivers/adodb-postgres64.inc.php	2006-02-16 00:20:48.000000000 +0100
@@ -464,6 +464,7 @@
 	{
 	global $ADODB_FETCH_MODE;
 	
+		$ADODB_FETCH_MODE=ADODB_FETCH_NUM;
 		$schema = false;
 		$false = false;
 		$this->_findschema($table,$schema);
