diff -urw adodb.481/adodb-lib.inc.php adodb/adodb-lib.inc.php
--- adodb.481/adodb-lib.inc.php	2006-05-03 09:42:47.000000000 +0200
+++ adodb/adodb-lib.inc.php	2006-05-15 13:52:14.000000000 +0200
@@ -783,7 +783,7 @@
 	$values = substr($values, 0, -2);
 
 	// Append the fields and their values to the insert query.
-	return 'INSERT INTO '.$tableName.' ( '.$fields.' ) VALUES ( '.$values.' )';
+	return 'INSERT INTO '.$zthis->nameQuote.$tableName.$zthis->nameQuote.' ( '.$fields.' ) VALUES ( '.$values.' )';
 }
 
 
diff -urw adodb.481/datadict/datadict-oci8.inc.php adodb/datadict/datadict-oci8.inc.php
--- adodb.481/datadict/datadict-oci8.inc.php	2006-05-03 09:42:28.000000000 +0200
+++ adodb/datadict/datadict-oci8.inc.php	2006-05-15 13:52:14.000000000 +0200
@@ -251,8 +251,8 @@
 		}
 		
 		if ( is_array($flds) )
-			$flds = implode(', ',$flds);
-		$s = 'CREATE' . $unique . ' INDEX ' . $idxname . ' ON ' . $tabname . ' (' . $flds . ')';
+			$flds = implode($this->quote.', '.$this->quote,$flds);
+		$s = 'CREATE' . $unique . ' INDEX ' .$this->quote. $idxname .$this->quote. ' ON ' . $this->quote.$tabname . $this->quote.' (' .$this->quote. $flds .$this->quote. ')';
 		
 		if ( isset($idxoptions[$this->upperName]) )
 			$s .= $idxoptions[$this->upperName];
diff -urw adodb.481/datadict/datadict-postgres.inc.php adodb/datadict/datadict-postgres.inc.php
--- adodb.481/datadict/datadict-postgres.inc.php	2006-05-03 09:42:28.000000000 +0200
+++ adodb/datadict/datadict-postgres.inc.php	2006-05-15 13:52:14.000000000 +0200
@@ -342,7 +342,7 @@
 		
 		$unique = isset($idxoptions['UNIQUE']) ? ' UNIQUE' : '';
 		
-		$s = 'CREATE' . $unique . ' INDEX ' . $idxname . ' ON ' . $tabname . ' ';
+		$s = 'CREATE' . $unique . ' INDEX "' . $idxname . '" ON "' . $tabname . '" ';
 		
 		if (isset($idxoptions['HASH']))
 			$s .= 'USING HASH ';
@@ -351,8 +351,8 @@
 			$s .= $idxoptions[$this->upperName];
 		
 		if ( is_array($flds) )
-			$flds = implode(', ',$flds);
-		$s .= '(' . $flds . ')';
+			$flds = implode('", "',$flds);
+		$s .= '("' . $flds . '")';
 		$sql[] = $s;
 		
 		return $sql;
diff -urw adodb.481/drivers/adodb-oci8.inc.php adodb/drivers/adodb-oci8.inc.php
--- adodb.481/drivers/adodb-oci8.inc.php	2006-05-03 09:42:44.000000000 +0200
+++ adodb/drivers/adodb-oci8.inc.php	2006-05-15 13:52:14.000000000 +0200
@@ -104,7 +104,7 @@
 		$ADODB_FETCH_MODE = ADODB_FETCH_NUM;
 		if ($this->fetchMode !== false) $savem = $this->SetFetchMode(false);
 		
-		$rs = $this->Execute(sprintf($this->metaColumnsSQL,strtoupper($table)));
+		$rs = $this->Execute(sprintf($this->metaColumnsSQL,$table));
 		
 		if (isset($savem)) $this->SetFetchMode($savem);
 		$ADODB_FETCH_MODE = $save;
