diff -urN adodb-494.orig/adodb-lib.inc.php adodb-494/adodb-lib.inc.php
--- adodb-494.orig/adodb-lib.inc.php	2007-02-26 12:51:09.000000000 +0100
+++ adodb-494/adodb-lib.inc.php	2007-02-26 12:51:28.000000000 +0100
@@ -857,7 +857,7 @@
 	$values = substr($values, 0, -2);
 
 	// Append the fields and their values to the insert query.
-	return 'INSERT INTO '.$tableName.' ( '.$fields.' ) VALUES ( '.$values.' )';
+	return 'INSERT INTO '.$zthis->nameQuote.$tableName.$zthis->nameQuote.' ( '.$fields.' ) VALUES ( '.$values.' )';
 }
 
 
diff -urN adodb-494.orig/datadict/datadict-oci8.inc.php adodb-494/datadict/datadict-oci8.inc.php
--- adodb-494.orig/datadict/datadict-oci8.inc.php	2007-02-26 12:51:09.000000000 +0100
+++ adodb-494/datadict/datadict-oci8.inc.php	2007-02-26 12:51:28.000000000 +0100
@@ -109,6 +109,7 @@
 	
 	function AddColumnSQL($tabname, $flds)
 	{
+		$tabname = $this->TableName ($tabname);
 		$f = array();
 		list($lines,$pkey) = $this->_GenFields($flds);
 		$s = "ALTER TABLE $tabname ADD (";
@@ -123,6 +124,7 @@
 	
 	function AlterColumnSQL($tabname, $flds)
 	{
+		$tabname = $this->TableName ($tabname);
 		$f = array();
 		list($lines,$pkey) = $this->_GenFields($flds);
 		$s = "ALTER TABLE $tabname MODIFY(";
@@ -136,6 +138,7 @@
 	
 	function DropColumnSQL($tabname, $flds)
 	{
+		$tabname = $this->TableName ($tabname);
 		if (!is_array($flds)) $flds = explode(',',$flds);
 		foreach ($flds as $k => $v) $flds[$k] = $this->NameQuote($v);
 		
@@ -259,9 +262,9 @@
 		}
 		
 		if ( is_array($flds) )
-			$flds = implode(', ',$flds);
-		$s = 'CREATE' . $unique . ' INDEX ' . $idxname . ' ON ' . $tabname . ' (' . $flds . ')';
-		
+			$flds = implode('", "',$flds);
+		$s = 'CREATE' . $unique . ' INDEX "' . $idxname . '" ON "' .$tabname . '" ("' . $flds . '")';
+
 		if ( isset($idxoptions[$this->upperName]) )
 			$s .= $idxoptions[$this->upperName];
 		
diff -urN adodb-494.orig/datadict/datadict-postgres.inc.php adodb-494/datadict/datadict-postgres.inc.php
--- adodb-494.orig/datadict/datadict-postgres.inc.php	2007-02-26 12:51:09.000000000 +0100
+++ adodb-494/datadict/datadict-postgres.inc.php	2007-02-26 12:51:28.000000000 +0100
@@ -342,7 +342,7 @@
 		
 		$unique = isset($idxoptions['UNIQUE']) ? ' UNIQUE' : '';
 		
-		$s = 'CREATE' . $unique . ' INDEX ' . $idxname . ' ON ' . $tabname . ' ';
+		$s = 'CREATE' . $unique . ' INDEX "' . $idxname . '" ON "' . $tabname . '" ';
 		
 		if (isset($idxoptions['HASH']))
 			$s .= 'USING HASH ';
@@ -351,8 +351,8 @@
 			$s .= $idxoptions[$this->upperName];
 		
 		if ( is_array($flds) )
-			$flds = implode(', ',$flds);
-		$s .= '(' . $flds . ')';
+			$flds = implode('", "',$flds);
+		$s .= '("' . $flds . '")';
 		$sql[] = $s;
 		
 		return $sql;
diff -urN adodb-494.orig/drivers/adodb-oci8.inc.php adodb-494/drivers/adodb-oci8.inc.php
--- adodb-494.orig/drivers/adodb-oci8.inc.php	2007-02-26 12:51:09.000000000 +0100
+++ adodb-494/drivers/adodb-oci8.inc.php	2007-02-26 12:51:28.000000000 +0100
@@ -105,7 +105,7 @@
 		$ADODB_FETCH_MODE = ADODB_FETCH_NUM;
 		if ($this->fetchMode !== false) $savem = $this->SetFetchMode(false);
 		
-		$rs = $this->Execute(sprintf($this->metaColumnsSQL,strtoupper($table)));
+		$rs = $this->Execute(sprintf($this->metaColumnsSQL,$table));
 		
 		if (isset($savem)) $this->SetFetchMode($savem);
 		$ADODB_FETCH_MODE = $save;
