diff -urw adodb5/drivers/adodb-odbc.inc.php adodb/drivers/adodb-odbc.inc.php
--- adodb5/drivers/adodb-odbc.inc.php	2006-05-03 09:42:44.000000000 +0200
+++ adodb/drivers/adodb-odbc.inc.php	2006-05-15 13:52:08.000000000 +0200
@@ -513,6 +513,11 @@
 					$this->_errorMsg = odbc_errormsg();
 					$this->_errorCode = odbc_error();
 				}
+				if($this->_errorCode == '00000') {	// MS SQL Server sometimes returns this in combination with the FreeTDS
+					$this->_errorMsg = '';		// driver and UnixODBC under Linux. This fixes the bogus "error"
+					$this->_errorCode = 0;		// <karsten@typo3.org>
+					return true;
+				}
 				return false;
 			}
 		
@@ -524,6 +529,11 @@
 					$this->_errorMsg = odbc_errormsg();
 					$this->_errorCode = odbc_error();
 				}
+				if($this->_errorCode == '00000') {	// MS SQL Server sometimes returns this in combination with the FreeTDS
+					$this->_errorMsg = '';		// driver and UnixODBC under Linux. This fixes the bogus "error"
+					$this->_errorCode = 0;		// <karsten@typo3.org>
+					return true;
+				}
 				return false;
 			}
 		} else
